### PR TITLE
Major refactor of clinvar_raw to simplify code flow

### PR DESCRIPTION
**Pull request by theferrit32**
- Removes use of core.async channels and futures
- Converts some functions into pure functions with no side effects

### CODE REVIEW

**Comment by larrybabb** [comment]
@theferrit32 is there an issue on the github board to associate with this?

**Comment by theferrit32** [comment]
@larrybabb this was done as part of the work to ingest the larger test dataset. It significantly speeds things up, and makes it less error prone. I would like to clean this up before merging to master but it is not a high priority because it is not blocking anything else right now.

An issue could be added and put in the backlog.

File: src/clinvar_raw/core.clj [hunk 1]
```diff
@@ -12,23 +12,17 @@
...
-  (.start (Thread. (partial stream/process-drop-messages (cfg/app-config))))
-  (.start (Thread. (partial stream/send-producer-messages
-                            (cfg/app-config)
-                            (cfg/kafka-config (cfg/app-config)))))
+#_(defn -main-unsafe
+    [& args]
+    (.start (Thread. (partial stream/process-drop-messages (cfg/app-config))))
+    (.start (Thread. (partial stream/send-producer-messages
+                              (cfg/app-config)
+                              (cfg/kafka-config (cfg/app-config)))))
```
Comment (tbl3rd) [inline src/clinvar_raw/core.clj]: Why not just delete this?

File: src/clinvar_raw/stream.clj [hunk 2]
```diff
@@ -50,23 +45,25 @@
...
+  (log/infof "Sending message with key %s to topic %s" key topic)
   (jc/send! producer (jd/->ProducerRecord {:topic-name topic}
                                           key
                                           (if (string? value) value (json/generate-string value))))
   (swap! send-update-to-exchange-counter inc))
 
-(defn google-storage-line-reader [bucket filename]
+(defn google-storage-line-reader
   "Returns a reader to the storage object. Caller must open (with-open)"
+  [bucket filename]
```
Comment (tbl3rd) [inline src/clinvar_raw/stream.clj:57]: The docstring could explain bucket and filename.

File: src/clinvar_raw/stream.clj [hunk 3]
```diff
@@ -50,23 +45,25 @@
...
   (log/debugf "Opening gs://%s/%s" bucket filename)
   (let [blob-id (BlobId/of bucket filename)
         blob (.get gc-storage blob-id)]
     (log/debugf "Obtaining reader to blob %s/%s" bucket filename)
     (-> blob (.reader (make-array Blob$BlobSourceOption 0)) (Channels/newReader "UTF-8") BufferedReader.)))
 
-(defn line-map-to-event [line-map entity-type release_date event-type]
+(defn line-map-to-event
   "Parses a single line of a drop file, transforms into an event object map"
+  [line-map entity-type release_date event-type]
```
Comment (tbl3rd) [inline src/clinvar_raw/stream.clj:71]: Document the parameters?

File: src/clinvar_raw/stream.clj [hunk 4]
```diff
@@ -133,127 +107,109 @@
...
+(defn dissoc-nil-values-recur
+  "Removes keys from maps for which the value is nil. Recurses with clojure.walk/postwalk"
+  [input-map]
+  (postwalk #(if (map? %) (dissoc-nil-values %) %) input-map))
+
+(defn select-keys-if
+  "Same as select-keys, except only selects a given key k if the value of (get m k) is not nil"
+  [m keys]
+  (->> (select-keys m keys)
+       dissoc-nil-values))
```
Comment (tbl3rd) [inline src/clinvar_raw/stream.clj]: Where is this used?

File: src/clinvar_raw/stream.clj [hunk 5]
```diff
@@ -133,127 +107,109 @@
...
+                                                      (:event-type procedure)))
+                             doall))))))))
+
+      ; Output end sentinel
+      (create-sentinel-message release-date :end)])))
+
+(defn dissoc-nil-values
+  "Removes keys from maps for which the value is nil."
+  [input-map]
+  (apply dissoc input-map (filter #(= nil (get input-map %)) (keys input-map))))
```
Comment (tbl3rd) [inline src/clinvar_raw/stream.clj]:
```clojure
(defn dissoc-nil-values
  "Remove entries with nil values from `input` map."
  [input]
  (into {} (remove (comp nil? val) input)))
```

File: src/clinvar_raw/stream.clj [hunk 6]
```diff
@@ -133,127 +107,109 @@
...
+       (while @listening-for-drop
+         (let [msgs (jc/poll consumer 100)]
+           (doseq [m msgs]
+             (log/tracef "Received message: %s" m)
+             (let [output-messages (process-clinvar-drop (:value m)
+                                                         (select-keys opts [:storage-protocol]))]
+               (log/info "Finished processing clinvar drop")
+               (log/info "Sending messages to output topic")
+               (doseq [output-m output-messages]
+                 (send-update-to-exchange producer output-topic output-m))))))))))
```
Comment (tbl3rd) [inline src/clinvar_raw/stream.clj]: I left some comments on the /tbl/async-refactor/ branch.

**Review by larrybabb — APPROVED** [review]
As usual I will defer to the clojure experts for final say, but these refinements look sound to me. Pls don't delay in merging them if they are needed to get our urgent goals accomplished.

**Review by larrybabb — APPROVED** [review]

**Comment by theferrit32** [comment]
I've addressed feedback from Tom aside from the file processing. That should be revisited later to more match other patterns for large sequence or file processing in our code.
https://github.com/clingen-data-model/clinvar-streams/issues/63

**Pull request merged**